### PR TITLE
Added new unit tests for src/enqueue.cc

### PR DESCRIPTION
**Pull request by atulkulk**
**Work item:** _"Internal", or link to GitHub issue (if applicable)._
Added new unit tests to cover test cases that were not covered by the functional tests for src/enqueue.cc and src/enqueue.h files.

**What were the changes?**  
Addition of new unit tests.

**Why were the changes made?**  
The motivation behind this task was to improve the overall code quality.

**How was the outcome achieved?**  
| File           | State   | Function Coverage | Line Coverage   | Region Coverage | Branch Coverage |
|-----------------|---------|-------------------|-----------------|-----------------|-----------------|
| **enqueue.cc** | Initial | 85.71% (42/49)	|  69.78% (1388/1989)	  | 54.48% (1501/2755)	|  46.10% (792/1718) |
|                 | Updated | 85.71% (42/49)	|  70.52% (1402/1988)	|  56.04% (1544/2755)	|  47.38% (814/1718) |
| **enqueue.h** | Initial | 33.33% (1/3)	|  33.33% (3/9)	|  42.86% (6/14)	|  50.00% (4/8) |
|                 | Updated | 100.00% (3/3)	|  100.00% (9/9)	|  100.00% (14/14)	|  100.00% (8/8) |

**Additional Documentation:**  
_What else should the reviewer know?_

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

File: test/EnqueueTests.cpp [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    size_t maxStackSize = 0;
+    ncclResult_t result = ncclInitKernelsForDevice(906, 65536, &maxStackSize);
+
+    EXPECT_EQ(result, ncclSuccess);
+    EXPECT_GT(maxStackSize, 0); // Should be set to a positive value
+}
+
+TEST_F(EnqueueTests, ncclInitKernelsForDevice_InvalidArch) {
+    size_t maxStackSize = 0;
+    ncclResult_t result = ncclInitKernelsForDevice(-1, 65536, &maxStackSize);
```
Comment (rahulvaidya20) [inline test/EnqueueTests.cpp:141]: Does it return `ncclSuccess` despite setting `cudaArch=-1`?

**Review by rahulvaidya20 — APPROVED** [review]
LGTM overall. Although I feel we could add more tests to cover edge cases for certain functions.

File: test/EnqueueTests.cpp (hunk 1 above)
Comment (atulkulk) [reply test/EnqueueTests.cpp:141]: Yes, it does return ncclSuccess.

**Comment by venksubramd** [comment]
Looks like
TEST_F(EnqueueTests, ncclInitKernelsForDevice_ValidInput) {
and
TEST_F(EnqueueTests, ncclInitKernelsForDevice_ValidStackSizePointer) {
are doing the same?

If so, we could remove one of them and update the other to indicate all parameters are expected to be valid in this test.

Do we need tests for other functions in enqueue.h, like ncclLaunchKernelBefore_NoUncapturedCuda, for example?

**Comment by rahulvaidya20** [comment]
> Looks like TEST_F(EnqueueTests, ncclInitKernelsForDevice_ValidInput) { and TEST_F(EnqueueTests, ncclInitKernelsForDevice_ValidStackSizePointer) { are doing the same?
> 
One of them can be removed I think.

> If so, we could remove one of them and update the other to indicate all parameters are expected to be valid in this test.
> 
> Do we need tests for other functions in enqueue.h, like ncclLaunchKernelBefore_NoUncapturedCuda, for example?

Those might already be covered through other forms of tests, I believe, as the code coverage for this file is 100%.

File: test/EnqueueTests.cpp [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        size_t bufferSize = 1024 * sizeof(float);
+        hipError_t hipErr = hipMalloc(&sendbuff, bufferSize);
+        ASSERT_EQ(hipErr, hipSuccess) << "Failed to allocate sendbuff";
+
+        hipErr = hipMalloc(&recvbuff, bufferSize);
+        ASSERT_EQ(hipErr, hipSuccess) << "Failed to allocate recvbuff";
+
+        // Initialize communicator
+        comm = new ncclComm();
+        memset(comm, 0, sizeof(ncclComm));
```
Comment (PJAvinash) [inline test/EnqueueTests.cpp:35]: ncclComm* comm = new ncclComm{}; // for C struct should do , instead of explicit memset()

File: test/EnqueueTests.cpp [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+// Static member definitions
+uint32_t EnqueueTests::abortFlag0 = 0;
+uint32_t EnqueueTests::abortFlag1 = 0;
+int EnqueueTests::abortFlagRefCount = 0;
+
+// Test ncclInitKernelsForDevice function
+TEST_F(EnqueueTests, ncclInitKernelsForDevice_ValidInput) {
+    size_t maxStackSize = 0;
+    ncclResult_t result = ncclInitKernelsForDevice(906, 65536, &maxStackSize);
```
Comment (PJAvinash) [inline test/EnqueueTests.cpp:127]: ncclInitKernelsForDevice(906 /*Explain this number for clear documentation */, 65536, &maxStackSize);

File: test/EnqueueTests.cpp [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+    // Test with single rank
+    EXPECT_EQ(ncclFuncSendCount(ncclFuncAllReduce, nRanks, count), count);
+    EXPECT_EQ(ncclFuncRecvCount(ncclFuncAllReduce, nRanks, count), count);
+    EXPECT_EQ(ncclFuncMaxSendRecvCount(ncclFuncAllReduce, nRanks, count), count);
+}
+
+TEST_F(EnqueueTests, ncclFuncCounts_LargeRankCount) {
+    size_t count = 1000;
+    int nRanks = 1024;
```
Comment (PJAvinash) [inline test/EnqueueTests.cpp:316]: what happens if nRanks = some odd number otherthan 1 , 0 , or a negative input ?

**Review by PJAvinash — APPROVED** [review]

**Review by Nikhil-Nunna — APPROVED** [review]

**Comment by rahulvaidya20** [comment]
@PJAvinash, I'll create a ticket to address your comments and add additional tests.
